File: pipeline/src/bytebrains_pipeline/simulator.py
"""The simulator coefficient exporter (ticket T7).

The prescriptive half of ADR-0002: Python exports per-source-market yield
coefficients (from the `source_market` fragment) into a `simulator` fragment of
the data bundle; the browser recomputes receipts, yield per visitor and the
Missing Billions gap for any user-chosen market mix with EXACTLY the arithmetic
in `simulate_mix` — transparent constant-2019-prices arithmetic, no ML, no
network.

Arithmetic contract (mirrored 1:1 in `dashboard/src/lib/simulator.ts`):

    yield_per_visitor_real   = sum(share_i * yield_real_i)          [2019 RM]
    receipts_real            = visitors * yield_per_visitor_real    [2019 RM]
    counterfactual           = visitors * anchor_per_visitor_real   [2019 RM]
    gap                      = counterfactual - receipts_real       (positive = missing billions)

Reconciliation: the fragment's markets partition the national totals the
`missing_billions` fragment uses (TSA Jad 1A receipts, visitor-basis arrivals)
via an explicit "Other markets (residual)" row, so at the 2024 actual mix the
simulator equals the headline calculator exactly (float tolerance). The
residual absorbs (a) the non-top-20 markets, (b) the four partial-coverage
top-20 markets (Bangladesh, Myanmar = arrivals-only; Canada, Netherlands =
receipts-only), and (c) the In Brief vs Jad 1A receipts basis difference. All
three are stated in the fragment, never silent.

Deflator: per-market yields are deflated by the NATIONAL CPI — no market-level
price index exists. Stated in the fragment.
"""
from __future__ import annotations

from .bundle import (
    MacroSeries,
    MissingBillionsFragment,
    SimulatorFragment,
    SimulatorMarket,
    SourceMarketFragment,
    Series,
)

RESIDUAL_MARKET = "Other markets (residual)"
# RM million. The reconciliation is exact arithmetic on the same doubles
# (same additions in the same order); the tolerance only absorbs float
# non-associativity across the two implementations.
RECONCILIATION_TOLERANCE_RM_M = 1e-6


def simulate_mix(
    yields_real: list[float],
    shares: list[float],
    visitor_arrivals: int,
    anchor_per_visitor_real: float,
) -> dict:
    """The client-side arithmetic, defined ONCE here and mirrored in
    `dashboard/src/lib/simulator.ts`. Same operations, same order: plain
    left-to-right float sums, no rounding, no numpy."""
    if len(yields_real) != len(shares):
        raise ValueError("yields and shares must have the same length")
    if any(s < 0 for s in shares):
        raise ValueError("shares must be non-negative")
    yield_per_visitor = 0.0
    contributions: list[float] = []
    for share, y in zip(shares, yields_real):
        yield_per_visitor += share * y  # left-to-right, the order TS must copy
    for share, y in zip(shares, yields_real):
        contributions.append(visitor_arrivals * (share * y) / 1_000_000.0)
    receipts = visitor_arrivals * yield_per_visitor / 1_000_000.0
    counterfactual = visitor_arrivals * anchor_per_visitor_real / 1_000_000.0
    return {
        "yield_per_visitor_real_rm": yield_per_visitor,
        "receipts_2019_prices_rm_million": receipts,
        "counterfactual_2019_prices_rm_million": counterfactual,
        "gap_2019_prices_rm_million": counterfactual - receipts,
        "contributions_2019_prices_rm_million": contributions,
    }


def _year_row(fragment: MissingBillionsFragment, year: int):
    row = next((y for y in fragment.years if y.year == year), None)
    if row is None:
        raise ValueError(f"missing_billions fragment has no {year} row")
    return row


def build_simulator_fragment(
    source_market: SourceMarketFragment,
    missing_billions: MissingBillionsFragment,
) -> SimulatorFragment:
    """Export the simulator coefficients from already-validated fragments.

    Never re-extracts anything: the yields are the source_market fragment's
    own figures (the same numbers segmentation reconciles against), deflated
    by the missing_billions deflator.
    """
    anchor_row = _year_row(missing_billions, missing_billions.anchor_year)
    row24 = _year_row(missing_billions, 2024)
    row23 = _year_row(missing_billions, 2023)
    ratio = row24.cpi_ratio_to_anchor

    markets: list[SimulatorMarket] = []
    sum_arrivals_2024 = 0
    sum_arrivals_2023 = 0
    sum_real_receipts = 0.0  # RM million, 2019 prices, left-to-right

    for row in source_market.markets:
        obs24 = next(o for o in row.observations if o.year == 2024)
        obs23 = next(o for o in row.observations if o.year == 2023)
        if row.coverage != "both":
            # partial-coverage markets carry no (yield, arrivals) pair for the
            # mix arithmetic; they are absorbed by the residual, never zeroed
            continue
        nominal_yield = obs24.yield_rm_per_visitor
        assert nominal_yield is not None and obs24.arrivals_persons is not None
        assert obs23.arrivals_persons is not None
        real_yield = nominal_yield / ratio
        share24 = obs24.arrivals_persons / row24.visitor_arrivals
        share23 = obs23.arrivals_persons / row23.visitor_arrivals
        markets.append(
            SimulatorMarket(
                market=row.market,
                coverage="both",
                yield_2024_nominal_rm_per_visitor=nominal_yield,
                yield_2024_real_2019_rm_per_visitor=real_yield,
                arrivals_2024_persons=obs24.arrivals_persons,
                arrivals_2023_persons=obs23.arrivals_persons,
                share_of_arrivals_2024=share24,
                share_of_arrivals_2023=share23,
            )
        )
        sum_arrivals_2024 += obs24.arrivals_persons
        sum_arrivals_2023 += obs23.arrivals_persons
        sum_real_receipts += real_yield * obs24.arrivals_persons / 1_000_000.0

    # the residual partitions the national totals the headline calculator uses
    residual_arrivals_2024 = int(row24.visitor_arrivals) - sum_arrivals_2024
    residual_arrivals_2023 = int(row23.visitor_arrivals) - sum_arrivals_2023
    residual_real_receipts = row24.actual_receipts_2019_prices_rm_million - sum_real_receipts
    if residual_arrivals_2024 <= 0 or residual_arrivals_2023 <= 0:
        raise ValueError(
            f"simulator residual: arrivals went non-negative-unfriendly "
            f"({residual_arrivals_2024:,} / {residual_arrivals_2023:,}) — the market set no "
            "longer fits inside the national totals; re-derive the residual, do not fudge it"
        )
    if residual_real_receipts <= 0:
        raise ValueError(
            f"simulator residual: real receipts went negative ({residual_real_receipts:,.1f} "
            "RM m, 2019 prices) — the In Brief top-20 receipts now exceed the Jad 1A national "
            "total (basis drift); re-derive the residual, do not fudge it"
        )
    residual_yield = residual_real_receipts * 1_000_000.0 / residual_arrivals_2024
    markets.append(
        SimulatorMarket(
            market=RESIDUAL_MARKET,
            coverage="residual",
            yield_2024_nominal_rm_per_visitor=None,
            yield_2024_real_2019_rm_per_visitor=residual_yield,
            arrivals_2024_persons=residual_arrivals_2024,
            arrivals_2023_persons=residual_arrivals_2023,
            share_of_arrivals_2024=residual_arrivals_2024 / row24.visitor_arrivals,
            share_of_arrivals_2023=residual_arrivals_2023 / row23.visitor_arrivals,
        )
    )

    return SimulatorFragment(
        prices="constant_2019_rm",
        anchor_year=missing_billions.anchor_year,
        mix_year=2024,
        comparison_year=2023,
        cpi_series_id=missing_billions.cpi_series_id,
        cpi_ratio_to_anchor_mix_year=ratio,
        visitor_arrivals_2024=int(row24.visitor_arrivals),
        visitor_arrivals_2023=int(row23.visitor_arrivals),
        anchor_per_visitor_real_2019_rm=anchor_row.per_visitor_real_2019_rm,
        markets=markets,
    )


def check_simulator_reconciliation(bundle) -> list[str]:
    """Ticket T7 ground truth: at the 2024 actual mix the simulator must
    reproduce the missing_billions fragment exactly (float tolerance)."""
    sim = bundle.fragments.get("simulator")
    if sim is None:
        raise ValueError("ground-truth check failed: simulator fragment missing")
    mb = bundle.fragments["missing_billions"]
    row24 = _year_row(mb, 2024)
    result = simulate_mix(
        yields_real=[m.yield_2024_real_2019_rm_per_visitor for m in sim.markets],
        shares=[m.share_of_arrivals_2024 for m in sim.markets],
        visitor_arrivals=sim.visitor_arrivals_2024,
        anchor_per_visitor_real=sim.anchor_per_visitor_real_2019_rm,
    )
    for field, expected in (
        ("receipts_2019_prices_rm_million", row24.actual_receipts_2019_prices_rm_million),
        ("counterfactual_2019_prices_rm_million", row24.counterfactual_receipts_2019_prices_rm_million),
        ("gap_2019_prices_rm_million", row24.gap_2019_prices_rm_million),
    ):
        if abs(result[field] - expected) > RECONCILIATION_TOLERANCE_RM_M:
            raise ValueError(
                f"simulator reconciliation failed: {field} {result[field]} != "
                f"missing_billions {expected} (tolerance {RECONCILIATION_TOLERANCE_RM_M})"
            )
    yield_diff = abs(result["yield_per_visitor_real_rm"] - row24.per_visitor_real_2019_rm)
    if yield_diff > 1e-6:
        raise ValueError(
            f"simulator reconciliation failed: per-visitor yield differs by {yield_diff} RM"
        )
    return [
        f"OK  simulator: {len(sim.markets)} market coefficients (constant 2019 prices, "
        f"deflated by {sim.cpi_series_id}); at the 2024 actual mix the client-side "
        f"arithmetic reproduces the Missing Billions headline exactly "
        f"(gap {result['gap_2019_prices_rm_million']:+,.4f} RM m, 2019 prices)"
    ]

File: dashboard/src/lib/simulator.ts
/**
 * Market-mix simulator arithmetic (ticket T7) — the client side of ADR-0002.
 *
 * The pipeline exports per-source-market yield coefficients (constant 2019
 * prices) in the bundle's `simulator` fragment; this module recomputes total
 * receipts, yield per visitor and the Missing Billions gap for ANY user-chosen
 * mix, live in the browser. No network, no ML.
 *
 * ARITHMETIC CONTRACT — mirrored 1:1 from
 * pipeline/src/bytebrains_pipeline/simulator.py::simulate_mix. Same operations
 * in the same order (plain left-to-right float sums, no rounding, no library
 * math). The round-trip fixture (tests/fixtures/simulator-fixture.json) was
 * written by the Python side; the tests must reproduce it from inputs alone.
 *
 * ALL FIGURES ARE CONSTANT-2019-PRICES (deflated by the DOSM national CPI).
 * Sign convention: gap = counterfactual (2019 mix held at 2019 real yield)
 * minus actual; POSITIVE = missing billions.
 */

export interface SimulatorMarket {
  market: string;
  coverage: "both" | "residual";
  yield_2024_nominal_rm_per_visitor: number | null;
  yield_2024_real_2019_rm_per_visitor: number;
  arrivals_2024_persons: number;
  arrivals_2023_persons: number;
  share_of_arrivals_2024: number;
  share_of_arrivals_2023: number;
}

export interface SimulatorFragment {
  prices: "constant_2019_rm";
  anchor_year: number; // 2019 — price anchor AND counterfactual yield anchor
  mix_year: number; // 2024
  comparison_year: number; // 2023 — earliest observed mix
  cpi_series_id: string;
  cpi_ratio_to_anchor_mix_year: number;
  visitor_arrivals_2024: number;
  visitor_arrivals_2023: number;
  anchor_per_visitor_real_2019_rm: number;
  markets: SimulatorMarket[];
}

export interface SimResult {
  /** Mix-weighted per-visitor receipts, 2019 RM. */
  yield_per_visitor_real_rm: number;
  /** Total receipts at the chosen mix, RM million, 2019 prices. */
  receipts_2019_prices_rm_million: number;
  /** Receipts if per-visitor yield equalled the 2019 anchor, RM million. */
  counterfactual_2019_prices_rm_million: number;
  /** counterfactual - receipts; POSITIVE = missing billions. */
  gap_2019_prices_rm_million: number;
  /** Per-market receipts contributions, RM million, 2019 prices. */
  contributions_2019_prices_rm_million: number[];
}

/**
 * THE arithmetic. Mirrors simulator.py::simulate_mix exactly — do not
 * "improve" the summation (no reduce-order tricks, no Kahan): the round-trip
 * test compares this bit-for-bit against the Python result.
 */
export function simulateMix(
  yieldsReal: number[],
  shares: number[],
  visitorArrivals: number,
  anchorPerVisitorReal: number
): SimResult {
  if (yieldsReal.length !== shares.length) {
    throw new Error("yields and shares must have the same length");
  }
  if (shares.some((s) => s < 0)) {
    throw new Error("shares must be non-negative");
  }
  let yieldPerVisitor = 0.0;
  for (let i = 0; i < shares.length; i++) {
    yieldPerVisitor += shares[i] * yieldsReal[i]; // left-to-right, same as Python
  }
  const contributions: number[] = [];
  for (let i = 0; i < shares.length; i++) {
    contributions.push((visitorArrivals * (shares[i] * yieldsReal[i])) / 1_000_000.0);
  }
  const receipts = (visitorArrivals * yieldPerVisitor) / 1_000_000.0;
  const counterfactual = (visitorArrivals * anchorPerVisitorReal) / 1_000_000.0;
  return {
    yield_per_visitor_real_rm: yieldPerVisitor,
    receipts_2019_prices_rm_million: receipts,
    counterfactual_2019_prices_rm_million: counterfactual,
    gap_2019_prices_rm_million: counterfactual - receipts,
    contributions_2019_prices_rm_million: contributions,
  };
}

/** Convenience wrapper: run the mix held on a fragment's own volumes/anchor. */
export function simulateFragmentMix(
  frag: SimulatorFragment,
  shares: number[]
): SimResult {
  return simulateMix(
    frag.markets.map((m) => m.yield_2024_real_2019_rm_per_visitor),
    shares,
    frag.visitor_arrivals_2024,
    frag.anchor_per_visitor_real_2019_rm
  );
}

// ---------------------------------------------------------------------------
// Presets. A preset is a SHARES vector (sums to 1 by construction); the UI
// turns shares back into slider weights. All rules are deterministic and
// stated in the UI.
// ---------------------------------------------------------------------------

export function shares2024(frag: SimulatorFragment): number[] {
  return frag.markets.map((m) => m.share_of_arrivals_2024);
}

export function shares2023(frag: SimulatorFragment): number[] {
  return frag.markets.map((m) => m.share_of_arrivals_2023);
}

// Policy preset rule (mirrors make_simulator_fixture.py::policy_mix): move this
// fraction of Singapore's 2024 share to the top-K markets by real 2019-price
// yield (excluding Singapore and the residual bucket), redistributed pro-rata
// by the recipients' own 2024 shares.
const POLICY_FROM_MARKET = "Singapore";
const POLICY_MOVE_FRACTION = 0.5;
const POLICY_TOP_K = 3;

export function policySingaporeShift(frag: SimulatorFragment): number[] {
  const shares = shares2024(frag);
  const fromI = frag.markets.findIndex((m) => m.market === POLICY_FROM_MARKET);
  if (fromI < 0) return shares;
  const ranked = frag.markets
    .map((m, i) => ({ i, yield: m.yield_2024_real_2019_rm_per_visitor, cov: m.coverage }))
    .filter(({ i, cov }) => cov === "both" && i !== fromI)
    .sort((a, b) => b.yield - a.yield);
  const recipients = ranked.slice(0, POLICY_TOP_K).map((r) => r.i);
  const move = shares[fromI] * POLICY_MOVE_FRACTION;
  const pool = recipients.reduce((sum, i) => sum + shares[i], 0);
  const out = [...shares];
  out[fromI] = shares[fromI] - move;
  for (const i of recipients) {
    out[i] = shares[i] + move * (shares[i] / pool);
  }
  return out;
}

/**
 * Turn shares into slider weights (percent points). The UI sliders hold raw
 * weights; effective shares are weight / total weight. A mix with zero total
 * weight falls back to uniform shares (never NaN).
 */
export function normalizeShares(weights: number[]): number[] {
  const total = weights.reduce((a, b) => a + b, 0);
  if (total <= 0) return weights.map(() => 1 / weights.length);
  return weights.map((w) => w / total);
}
